lib/rspec/shell/expectations/stubbed_env.rb

```diff
@@ -28,22 +28,78 @@ def cleanup
         end
 
         def stub_command(command)
+          wrap_command command
           StubbedCommand.new command, @dir
         end
 
         def execute(command, env_vars = {})
-          Open3.capture3(env_vars, "#{env} #{command}")
+          full_command=wrap_execute(<<-multiline_script
+            #{env} source #{command} 2> #{wrapped_error_path}
+          multiline_script
+          )
+
+          Open3.capture3(env_vars, full_command)
         end
 
         def execute_function(script, command, env_vars = {})
-          Open3.capture3(env_vars, "bash -c 'source #{script} && #{env} #{command}'")
+          full_command=wrap_execute(<<-multiline_script
+            source #{script} 2> #{wrapped_error_path}
+            #{env} #{command}
+          multiline_script
+          )
+
+          Open3.capture3(env_vars, full_command)
+        end
+
+        def execute_inline(command_string, env_vars = {})
+          temp_command_path=Dir::Tmpname.make_tmpname("#{@dir}/inline-", nil)
+          File.write(temp_command_path, command_string)
+          execute(temp_command_path, env_vars)
         end
 
         private
 
+        def wrap_command(command)
+          command_path = File.join(@dir, command)
```

Comment:
Maybe have the name reflect this? I can't think of anything good right now, but instead of wrap_command maybe something like wrap_command_for_doing_all_the_overriding_stuff